### DISCUSSION TITLE
Recommendation on passing input values in

### DISCUSSION THREAD

**Discussion by 4l1fe** [zauberzeug/nicegui#76] [Q&A] [answered]
What is a recommended way of passing the input values into a handler?

Is it a callback style only?

```python

def handle_inputs(*inputs):
   print(inputs)


with ui.row():
   ui.input()
   ui.input()
   ui.input()
   ...
   ui.button(on_click=handle_inputs)  # I'd like to pass all the final input values in at once here
```

**Comment by falkoschindler** [comment, marked as answer]
Well, my recommendation depends. The most trivial way would be to pass all values:

```python
def handle_inputs(a, b, c):
    print(a, b, c)

a = ui.input()
b = ui.input()
c = ui.input()
ui.button(on_click=lambda: handle_inputs(a.value, b.value, c.value))
```

If the list of inputs gets too long, you could access the UI elements directly:

```python
def handle_more_inputs():
    print(d.value, e.value, f.value)

d = ui.input()
e = ui.input()
f = ui.input()
ui.button(on_click=handle_more_inputs)
```

But maybe it's better to pack all data into a model class:

```python
@dataclass
class Model:
    g: str = ''
    h: str = ''
    i: str = ''

def handle_model(model):
    print(model.g, model.h, model.i)

model = Model()
ui.input().bind_value_to(model, 'g')
ui.input().bind_value_to(model, 'h')
ui.input().bind_value_to(model, 'i')
ui.button(on_click=lambda: handle_model(model))
```

In the examples above I dropped the `ui.row` container because currently there is no way to use it container for, e.g., accessing all its input values at once. But maybe the container could provide the collected values of all its input elements... This could look something like this:

```python
def handle_form(x, y, z):
    print(x, y, z)

with ui.row() as form:
    ui.input()
    ui.input()
    ui.input()
    ui.button(on_click=lambda: handle_form(*form.get_input_values()))
```

What do you think? Did you have something completely different in mind?

**Reply by 4l1fe** [reply]
All your examples are better than my current approach 🙂 ,due to lack of documentation, my eyes passed attr `input.value` and method `input.bind_value_to()`, so i'm getting back to them.

The model class or dataclass container is what i'm using and grouping the ui elements with a context manager to  get all their values looks really good as on a page many different controls could be placed. It simplifies subsequent validation and input filtering in the case of many controls per page. And i would give it a little more abstract name `group` instead of `form`, or even left it as is `row`. `form` is a pure HTTP term and ,if understand correctly, Nicegui hides an HTTP layer. But that's not important nuance. Grouping is nice to have , no doubt.
